README.md
<p align="center">
  <img src="docs/node-uri.png" alt="node-uri"/>
<p>

<p align="center">
  An <a href="https://tools.ietf.org/html/rfc3986" target="_blank">RFC-3986</a> compliant and zero-dependencies Node.js module to parse URIs, punycode, punydecode, test URIs, URLs, Sitemap URLs, domains, IPs but also encode and decode URIs, URLs and Sitemap URLs.
<p>

# Table of Contents

- [Table of Contents](#table-of-contents)
- [Presentation](#presentation)
- [Installation](#installation)
- [Technical information](#technical-information)
  - [Node.js](#nodejs)
  - [Tests](#tests)
    - [Linting](#linting)
    - [Unit](#unit)
- [Usage](#usage)
  - [Import](#import)
  - [punycode(domain)](#punycodedomain)
  - [punydecode(domain)](#punydecodedomain)
  - [parseURI(uri)](#parseuriuri)
  - [recomposeURI(components)](#recomposeuricomponents)
  - [isDomainLabel(label)](#isdomainlabellabel)
  - [isDomain(name)](#isdomainname)
  - [isIP(ip)](#isipip)
  - [isIPv4(ip)](#isipv4ip)
  - [isIPv6(ip)](#isipv6ip)
  - [checkURI(uri)](#checkuriuri)
  - [checkHttpURL(uri)](#checkhttpurluri)
  - [checkHttpsURL(uri)](#checkhttpsurluri)
  - [checkHttpSitemapURL(uri)](#checkhttpsitemapurluri)
  - [checkHttpsSitemapURL(uri)](#checkhttpssitemapurluri)
  - [checkWebURL(uri)](#checkweburluri)
  - [checkSitemapURL(uri)](#checksitemapurluri)
  - [encodeURIComponentString(component, options)](#encodeuricomponentstringcomponent-options)
  - [encodeURIString(uri, options)](#encodeuristringuri-options)
  - [encodeWebURL(uri, options)](#encodeweburluri-options)
  - [encodeSitemapURL(uri)](#encodesitemapurluri)
  - [decodeURIComponentString(component, options)](#decodeuricomponentstringcomponent-options)
  - [decodeURIString(uri, options)](#decodeuristringuri-options)
  - [decodeWebURL(uri, options)](#decodeweburluri-options)
  - [decodeSitemapURL(uri, options)](#decodesitemapurluri-options)
  - [Errors](#errors)
    - [Object structure](#object-structure)
    - [Codes](#codes)
- [Code of Conduct](#code-of-conduct)
- [Contributing](#contributing)
- [Support](#support)
- [Security](#security)
- [License](#license)

# Presentation

Few libraries exist in the Node.js sphere providing helpers to deal with URIs and to be fully **RFC-3986** compliant without any dependency.

Amongst being **RFC-3986** compliant and having no dependency, *node-uri* aims to support Sitemap URLs for SEO purposes, fix `url.parse` that automatically punycodes a host and cannot parse URIs other than URLs, fix `encodeURI` and `encodeURIComponent` native functions that relies on an old standard prior to **RFC-3986** and fix `decodeURI` that does not support IDNs and cannot properly work with `encodeURI` since the function is based on an outdated standard.

The main features of this project are:

- parse any URI (URNs, URLs, URIs with IDNs, etc.);
- get the safe Punycode ASCII or Unicode serialization of a domain;
- check an URI, HTTP/HTTPS/Sitemap URL, IP, domain is valid with clear checking errors;
- encode/decode an URI, HTTP/HTTPS/Sitemap URL.

**2838 assertions** ensure parsing, encoding, decoding, checking URIs, URLs, IPs, domains are working as expected. This does not make this library a 100% reliable source. So if you find any errors, please feel free to report, [contribute](#contributing) and help fixing any issues.

To make sure we properly understand the difference between an URI and an URL, these two are not exactly the same thing. An URI is an identifier of a specific resource. Like a page, a book, or a document. An URL is a special type of identifier that also tells you how to access it, such as HTTPs, FTP, etc. If the protocol (https, ftp, etc.) is either present or implied for a domain, you should call it an URL even though it’s also an URI.

# Installation

`npm install node-uri`

`npm i -S node-uri`

# Technical information

## Node.js

- Language: JavaScript ES6/ES7
- VM: Node.js >= Carbon (8.17.0)

## Tests

Node.js >= Dubnium (10.22.1) could be required for some testing modules.

### Linting

ESLint with Airbnb base rules. See  __<a href="https://github.com/airbnb/javascript" target="_blank">Airbnb JavaScript Style Guide</a>__.

`npm run lint`

### Unit

Mocha and Chai.

`npm test`

# Usage

## Import

```javascript
const uri = require('node-uri');

// uri is an object of functions
const {
  punycode,
  punydecode,
  parseURI,
  recomposeURI,
  isDomainLabel,
  isDomain,
  isIP,
  isIPv4,
  isIPv6,
  checkURI,
  checkHttpURL,
  checkHttpsURL,
  checkHttpSitemapURL,
  checkHttpsSitemapURL,
  checkWebURL,
  checkSitemapURL,
  encodeURIComponentString,
  encodeURIString,
  encodeWebURL,
  encodeSitemapURL,
  decodeURIComponentString,
  decodeURIString,
  decodeWebURL,
  decodeSitemapURL,
} = require('node-uri');
```

*node-uri* module exports an object of functions. You'll find the complete list of functions below.

- `uri` **<Object\>** with the following functions.

## punycode(domain)

Returns the Punycode ASCII serialization of the domain. If domain is an invalid domain, the empty string is returned.

**Note**:

- native function `url.domainToASCII` does not support IPv6 only IPv4;
- native function `url.domainToASCII` throws if no domain is provided or returns `null`, `undefined`, `nan` for `null`, `undefined` or `NaN` values which is not what to be expected.

<br/>

- `domain` **<String\>**
- Returns: **<String\>**

<br/>

**Examples**:

```javascript
punycode(); // ''
punycode('a.b.c.d.e.fg'); // 'a.b.c.d.e.fg'
punycode('xn--iñvalid.com'); // ''
punycode('中文.com'); // 'xn--fiq228c.com'
punycode('xn--fiq228c.com'); // 'xn--fiq228c.com'
punycode('2001:db8:85a3:8d3:1319:8a2e:370:7348'); // '2001:db8:85a3:8d3:1319:8a2e:370:7348'
punycode('127.0.0.1'); // '127.0.0.1'
punycode(undefined|null|NaN); // ''
```

## punydecode(domain)

Returns the Unicode serialization of the domain. If domain is an invalid domain, the empty string is returned.

**Note**:

- native function `url.domainToUnicode` does not support IPv6 only IPv4;
- native function `url.domainToUnicode` throws if no domain is provided or returns `null`, `undefined`, `nan` for `null`, `undefined` or `NaN` values which is not what to be expected.

<br/>

- `domain` **<String\>**
- Returns: **<String\>**

<br/>

**Examples**:

```javascript
punydecode(); // ''
punydecode('a.b.c.d.e.fg'); // 'a.b.c.d.e.fg'
punydecode('xn--iñvalid.com'); // ''
punydecode('xn--fiq228c.com'); // '中文.com'
punydecode('中文.com'); // '中文.com'
punydecode('2001:db8:85a3:8d3:1319:8a2e:370:7348'); // '2001:db8:85a3:8d3:1319:8a2e:370:7348'
punydecode('127.0.0.1'); // '127.0.0.1'
punydecode(undefined|null|NaN); // ''
```

## parseURI(uri)

Parse a string to get URI components.

**Support**:

- IPv4 and IPv6 hosts;
- Internationalized Domain Name (IDN).

**Note**:

- RegExp from __<a href="https://tools.ietf.org/html/rfc3986#appendix-B" target="_blank">RFC-3986</a>__;
- scheme and host strings will always be put in lowercase once parsed, as specified in **RFC-3986**;
- authority and its components will be put at null values if authority parsed is missing or empty;
- **prefer using [checkURI](#checkuriuri) to parse and fully check an URI**.

<br/>

**Generic syntax**:

![URI Syntax](docs/uri-syntax.png "URI Syntax")

<br/>

**Example URIs**:

![RFC-3986](docs/rfc-3986.png "RFC-3986")

<br/>

**Based on**:

- __<a href="https://tools.ietf.org/html/rfc3986" target="_blank">RFC-3986</a>__.

<br/>

- `uri` **<String\>**
- Returns: **<Object\>**
  - `scheme` **<String\>** The URI scheme. *Default*: `null`
  - `authority` **<String\>** The URI authority with the Punycode ASCII serialization of the domain. *Default*: `null`
  - `authorityPunydecoded` **<String\>** The URI authority with the Unicode serialization of the domain. *Default*: `null`
  - `userinfo` **<String\>** The URI userinfo. *Default*: `null`
  - `host` **<String\>** The URI authority's host with the Punycode ASCII serialization of the domain. *Default*: `null`
  - `hostPunydecoded` **<String\>** The URI authority's host with the Unicode serialization of the domain. *Default*: `null`
  - `port` **<Number\>** || **<String\>** The URI authority's port. A string if not able to be parsed in an integer. *Default*: `null`
  - `path` **<String\>** The URI path. *Default*: `null`
  - `pathqf` **<String\>** The URI path, query and fragment. *Default*: `null`
  - `query` **<String\>** The URI query. *Default*: `null`
  - `fragment` **<String\>** The URI fragment. *Default*: `null`
  - `href` **<String\>** The URI recomposed. See __[recomposeURI](#recomposeuricomponents)__. *Default*: `null`

<br/>

**Examples**:

```javascript
parseURI();
// {
//   scheme: null,
//   authority: null,
//   authorityPunydecoded: null,
//   userinfo: null,
//   host: null,
//   hostPunydecoded: null,
//   port: null,
//   path: null,
//   pathqf: null,
//   query: null,
//   fragment: null,
//   href: null,
// }

parseURI('foo://user:pass@xn--fiq228c.com:8042/over/there?name=ferret#nose');
// {
//   scheme: 'foo',
//   authority: 'user:pass@xn--fiq228c.com:8042',
//   authorityPunydecoded: 'user:pass@中文.com:8042',
//   userinfo: 'user:pass',
//   host: 'xn--fiq228c.com',
//   hostPunydecoded: '中文.com',
//   port: 8042,
//   path: '/over/there',
//   pathqf: '/over/there?name=ferret#nose',
//   query: 'name=ferret',
//   fragment: 'nose',
//   href: 'foo://user:pass@xn--fiq228c.com:8042/over/there?name=ferret#nose',
// }

parseURI('foo://user:pass@中文.com:80g42/over/there?name=ferret#nose');
// {
//   scheme: 'foo',
//   authority: 'user:pass@xn--fiq228c.com:80g42',
//   authorityPunydecoded: 'user:pass@中文.com:80g42',
//   userinfo: 'user:pass',
//   host: 'xn--fiq228c.com',
//   hostPunydecoded: '中文.com',
//   port: '80g42',
//   path: '/over/there',
//   pathqf: '/over/there?name=ferret#nose',
//   query: 'name=ferret',
//   fragment: 'nose',
//   href: 'foo://user:pass@xn--fiq228c.com:80g42/over/there?name=ferret#nose',
// }

parseURI('urn:isbn:0-486-27557-4');
// {
//   scheme: 'urn',
//   authority: null,
//   authorityPunydecoded: null,
//   userinfo: null,
//   host: null,
//   hostPunydecoded: null,
//   port: null,
//   path: 'isbn:0-486-27557-4',
//   pathqf: 'isbn:0-486-27557-4',
//   query: null,
//   fragment: null
//   href: 'urn:isbn:0-486-27557-4',
// }

parseURI('http://user:pass@[fe80::7:8%eth0]:8080');
// {
//   scheme: 'http',
//   authority: 'user:pass@[fe80::7:8%eth0]:8080',
//   authorityPunydecoded: 'user:pass@[fe80::7:8%eth0]:8080',
//   userinfo: 'user:pass',
//   host: 'fe80::7:8%eth0',
//   hostPunydecoded: 'fe80::7:8%eth0',
//   port: 8080,
//   path: '',
//   pathqf: '',
//   query: null,
//   fragment: null,
//   href: 'http://user:pass@[fe80::7:8%eth0]:8080/'
// }
```

## recomposeURI(components)

Recompose an URI from its components with basic URI checking.

The empty string is returned if unable to recompose the URI.

**Rules**:

1. scheme is required and must be at least 1 character;
2. path is required and can be empty;
3. if host is present path must be empty or start with `/`;
4. if host is not present path must not start with `//`;
5. host, if any, must be at least 3 characters;
6. userinfo will be ignored if empty;
7. port will be ignored if unable to parse it into an integer between 0 - 65535;
8. query will be ignored if empty;
9. fragment will be ignored if empty.

**Support**:

- IPv4 and IPv6.

**Note**:

- `/` is added to any URI with a host and an empty path.

<br/>

**Generic syntax**:

![URI Syntax](docs/uri-syntax.png "URI Syntax")

<br/>

**Based on**:

- __<a href="https://tools.ietf.org/html/rfc3986" target="_blank">RFC-3986</a>__.

<br/>

- `components` **<Object\>**:
  - `scheme`* **<String\>** The URI scheme.
  - `userinfo` **<String\>** The URI userinfo.
  - `host` **<String\>** The URI authority's host.
  - `port` **<Number\>** The URI authority's port.
  - `path`* **<String\>** The URI path.
  - `query` **<String\>** The URI query.
  - `fragment` **<String\>** The URI fragment.
- Returns: **<String\>**

<br/>

**Examples**:

```javascript
recomposeURI(); // ''

recomposeURI({
  scheme: null,
  userinfo: 'user:pass',
  host: 'example.com',
  port: 8080,
  path: null,
  query: 'a=b',
  fragment: 'anchor',
}); // ''

recomposeURI({
  scheme: 'foo',
  userinfo: null,
  host: null,
  port: null,
  path: '',
  query: null,
  fragment: null,
}); // 'foo:'

recomposeURI({
  scheme: 'foo',
  userinfo: 'user:pass',
  host: 'bar.com',
  port: 8080,
  path: '/over/there',
  query: 'a=b',
  fragment: 'anchor',
}); // 'foo://user:pass@bar.com:8080/over/there?a=b#anchor'

recomposeURI({
  scheme: 'foo',
  userinfo: 'user:pass',
  host: 'fe80::7:8%eth0',
  port: '8080',
  path: '/over/there',
  query: 'a=b',
  fragment: 'anchor',
}); // 'foo://user:pass@[fe80::7:8%eth0]:8080/over/there?a=b#anchor'

recomposeURI({
  scheme: 'foo',
  userinfo: '',
  host: 'fe80::7:8%eth0',
  port: '55g55',
  path: '/over/there',
  query: '',
  fragment: '',
}); // 'foo://[fe80::7:8%eth0]/over/there'
```

## isDomainLabel(label)

Test a label is a valid domain label according to **RFC-1034**.

> "Note that while upper and lower case letters are allowed in domain names, no significance is attached to the case.  That is, two names with the same spelling but different case are to be treated as if identical."

By convention uppercased domain label will be considered invalid.

**Rules**:

1. "*Labels must be 63 characters or less.*";
2. can be minimum one character;
3. must only use lowercase letters, digits or hyphens;
4. must not start or end with a hyphen;
5. must not have consecutive hyphens;
6. can start or end with a digit.

<br/>

**Based on**:

- __<a href="https://www.ietf.org/rfc/rfc1034.txt" target="_blank">RFC-1034</a>__.

<br/>

- `label` **<String\>**
- Returns: **<Boolean\>**

<br/>

**Examples**:

```javascript
isDomainLabel('a'); // true
isDomainLabel('1a3'); // true
isDomainLabel('1-3'); // true
isDomainLabel('1-y'); // true

isDomainLabel(); // false
isDomainLabel('a'.repeat(64)); // false
isDomainLabel('A'); // false
isDomainLabel('-a'); // false
isDomainLabel('a-'); // false
isDomainLabel('-a'); // false
isDomainLabel('la--bel'); // false
isDomainLabel(undefined|null|NaN); // false
```

## isDomain(name)

Test a name is a valid domain according to **RFC-1034**.

Supports Fully-Qualified Domain Name (FQDN) and Internationalized Domain Name (IDN).

**Rules**:

1. __[labels rules apply](#isdomainlabellabel)__;
2. "*[...] the total number of octets that represent a domain name (i.e., the sum of all label octets and label lengths) is limited to 255.*";
3. labels are separated by dots (".");
4. must have at least one extension label;
5. must have labels different from each other;
6. last label can be empty (root label ".");
7. labels can start with `xn--` for IDNs if the ASCII serialization is a valid Punycode **and has valid characters**.

<br/>

**Based on**:

- __<a href="https://www.ietf.org/rfc/rfc1034.txt" target="_blank">RFC-1034</a>__.

<br/>

- `name` **<String\>**
- Returns: **<Boolean\>**

<br/>

**Examples**:

```javascript
isDomain('a.b'); // true
isDomain('a.b.'); // true
isDomain('中文.com'); // true
isDomain('xn--fiq228c.com'); // true
isDomain('www.中文.com'); // true
isDomain(`${'a'.repeat(63)}.${'b'.repeat(63)}.${'c'.repeat(63)}.${'d'.repeat(63)}`); // true

isDomain(); // false
isDomain('a'); // false
isDomain('a.a'); // false
isDomain('a.b.a'); // false
isDomain('a.b.a'); // false
isDomain('中文.xn--fiq228c.com'); // false
isDomain('www.xn--hf.com'); // false
isDomain(`${'a'.repeat(63)}.${'b'.repeat(63)}.${'c'.repeat(63)}.${'d'.repeat(63)}.`); // false
isDomain('xn--\'-6xd.com') // false even though xn--'-6xd is a valid Punycode for ॐ but has an invalid character
```

## isIP(ip)

Test a string is a valid IP.

Supports IPv4 and IPv6.

<br/>

- `ip` **<String\>**
- Returns: **<Boolean\>**

<br/>

**Examples**:

```javascript
isIP('23.71.254.72'); // true
isIP('1:2:3:4::6:7:8'); // true

isIP(); // false
isIP('100..100.100.100.'); // false
isIP('3ffe:b00::1::a'); // false
```

## isIPv4(ip)

Test a string is a valid IPv4.

<br/>

- `ip` **<String\>**
- Returns: **<Boolean\>**

<br/>

**Examples**:

```javascript
isIPv4('8.8.8.8'); // true
isIPv4('1:2::8'); // false
isIPv4(); // false
```

## isIPv6(ip)

Test a string is a valid IPv6.

<br/>

- `ip` **<String\>**
- Returns: **<Boolean\>**

<br/>

**Examples**:

```javascript
isIPv6('2001:0000:1234:0000:0000:C1C0:ABCD:0876'); // true
isIPv6('212.58.241.131'); // false
isIPv6(); // false
```

## checkURI(uri)

Check an URI is valid according to **RFC-3986**.

**Rules**:

1. scheme is required and cannot be empty;
2. path is required and can be empty;
3. if authority is present path must be empty or start with `/`;
4. if authority is not present path must not start with `//`;
5. __<a href="https://tools.ietf.org/html/rfc3986#section-3.1" target="_blank">scheme can only have specific characters</a>__;
6. if authority is present:

- host must be a valid IP or domain name;
- __<a href="https://tools.ietf.org/html/rfc3986#section-3.2.1" target="_blank">userinfo, if any, can only have specific characters</a>__;
- port, if any, must be an integer between 0 - 65535.

7. __<a href="https://tools.ietf.org/html/rfc3986#section-3.3" target="_blank">path, query and fragment can only have specific characters</a>__.

<br/>

**Generic syntax**:

![URI Syntax](docs/uri-syntax.png "URI Syntax")

<br/>

**Based on**:

- __<a href="https://tools.ietf.org/html/rfc3986" target="_blank">RFC-3986</a>__.

<br/>

- `uri` **<String\>**
- Returns: **<Object\>**
  - `scheme` **<String\>** The URI scheme.
  - `authority` **<String\>** The URI authority with the Punycode ASCII serialization of the domain. *Default*: `null`
  - `authorityPunydecoded` **<String\>** The URI authority with the Unicode serialization of the domain. *Default*: `null`
  - `userinfo` **<String\>** The URI userinfo. *Default*: `null`
  - `host` **<String\>** The URI authority's host with the Punycode ASCII serialization of the domain. *Default*: `null`
  - `hostPunydecoded` **<String\>** The URI authority's host with the Unicode serialization of the domain. *Default*: `null`
  - `port` **<Number\>** || **<String\>** The URI authority's port. A string if not able to be parsed in an integer. *Default*: `null`
  - `path` **<String\>** The URI path.
  - `pathqf` **<String\>** The URI path, query and fragment.
  - `query` **<String\>** The URI query. *Default*: `null`
  - `fragment` **<String\>** The URI fragment. *Default*: `null`
  - `href` **<String\>** The URI recomposed. *Default*: `null`
  - `valid` **<Boolean\>** Whether the URI is valid. *Default*: `false`
- Throws: **<URIError\>** If no error is thrown then the URI is valid. Error codes:
  - `URI_INVALID_TYPE`
  - `URI_MISSING_SCHEME`
  - `URI_EMPTY_SCHEME`
  - `URI_MISSING_PATH`
  - `URI_INVALID_PATH`
  - `URI_INVALID_HOST`
  - `URI_INVALID_SCHEME_CHAR`
  - `URI_INVALID_USERINFO_CHAR`
  - `URI_INVALID_PORT`
  - `URI_INVALID_PATH_CHAR`
  - `URI_INVALID_QUERY_CHAR`
  - `URI_INVALID_FRAGMENT_CHAR`
  - `URI_INVALID_PERCENT_ENCODING`

<br/>

**Examples**:

```javascript
checkURI(); // throws URIError with code URI_INVALID_TYPE
checkURI('://example.com'); // throws URIError with code URI_MISSING_SCHEME
checkURI('foo:////bar'); // throws URIError with code URI_INVALID_PATH
checkURI('foo://xn--iñvalid.com'); // throws URIError with code URI_INVALID_HOST
checkURI('fôo:bar'); // throws URIError with code URI_INVALID_SCHEME_CHAR
checkURI('foo://üser:pass@bar.com'); // throws URIError with code URI_INVALID_USERINFO_CHAR
checkURI('foo://bar.com:80g80'); // throws URIError with code URI_INVALID_PORT
checkURI('foo://bar.com/°'); // throws URIError with code URI_INVALID_PATH_CHAR
checkURI('foo://bar.com/over/there?quêry=5'); // throws URIError with code URI_INVALID_QUERY_CHAR
checkURI('foo://bar.com/over/there?query=5#anch#r'); // throws URIError with code URI_INVALID_FRAGMENT_CHAR
checkURI('http://www.bar.baz/foo%2') // throws URIError with code URI_INVALID_PERCENT_ENCODING

checkURI('foo://user:pass@xn--fiq228c.com:8042/over/there?name=ferret#nose');
// {
//   scheme: 'foo',
//   authority: 'user:pass@xn--fiq228c.com:8042',
//   authorityPunydecoded: 'user:pass@中文.com:8042',
//   userinfo: 'user:pass',
//   host: 'xn--fiq228c.com',
//   hostPunydecoded: '中文.com',
//   port: 8042,
//   path: '/over/there',
//   pathqf: '/over/there?name=ferret#nose',
//   query: 'name=ferret',
//   fragment: 'nose',
//   href: 'foo://user:pass@xn--fiq228c.com:8042/over/there?name=ferret#nose',
//   valid: true
// }
```

## checkHttpURL(uri)

Check an URI is a valid HTTP URL.

**Rules**:

1. __[must be a valid URI](#checkuriuri)__;
1. scheme must be `http` or `HTTP`;
2. authority is required;
3. URL must be less than 2048 characters.

<br/>

**Based on**:

- __<a href="https://tools.ietf.org/html/rfc3986" target="_blank">RFC-3986</a>__.

<br/>

- `uri` **<String\>**
- Returns: **<Object\>**
  - `scheme` **<String\>** The URL scheme.
  - `authority` **<String\>** The URL authority with the Punycode ASCII serialization of the domain. *Default*: `null`
  - `authorityPunydecoded` **<String\>** The URL authority with the Unicode serialization of the domain. *Default*: `null`
  - `userinfo` **<String\>** The URL userinfo. *Default*: `null`
  - `host` **<String\>** The URL authority's host with the Punycode ASCII serialization of the domain. *Default*: `null`
  - `hostPunydecoded` **<String\>** The URL authority's host with the Unicode serialization of the domain. *Default*: `null`
  - `port` **<Number\>** || **<String\>** The URL authority's port. A string if not able to be parsed in an integer. *Default*: `null`
  - `path` **<String\>** The URL path.
  - `pathqf` **<String\>** The URI path, query and fragment.
  - `query` **<String\>** The URL query. *Default*: `null`
  - `fragment` **<String\>** The URL fragment. *Default*: `null`
  - `href` **<String\>** The URL recomposed. *Default*: `null`
  - `valid` **<Boolean\>** Whether the URL is valid. *Default*: `false`
- Throws: **<URIError\>** If no error is thrown then the URL is valid. Error codes:
  - `URI_INVALID_TYPE`
  - `URI_MISSING_SCHEME`
  - `URI_EMPTY_SCHEME`
  - `URI_MISSING_PATH`
  - `URI_INVALID_PATH`
  - `URI_INVALID_HOST`
  - `URI_INVALID_USERINFO_CHAR`
  - `URI_INVALID_PORT`
  - `URI_INVALID_PATH_CHAR`

CHANGELOG.md
# CHANGELOG

## 1.1.2 - delivery @06/06/2023

- chore: update dev dependencies
- docs: update

## 1.1.1 - delivery @07/10/2020

- fix readme typo

## 1.1.0 - delivery @07/10/2020

- fix parser to support IPv6 extraction (parse and recompose)
- add href attribute when parsing URI
- fix punycode and punydecode that should work with IPv6
- fix " (quote) character not allowed in any URI
- fix sitemap characters
- remove checkURISyntax from library
- change recomposeURI to ignore port, userinfo, query and fragment if not valid
- add max URL length at 2048
- add integer range for port
- userinfo chars accept A-Z
- add and update tests
- update readme

## 1.0.1 - delivery @22/09/2020

- docs:
  - update project description
  - add missing github community files

## 1.0.0 - delivery @21/09/2020

- first delivery


lib/checkers/chars.js
/**
 * chars checkers
 *
 * - isSchemeChar(char, { start } = {}) -> Boolean
 * - isUserinfoChar(char, encode) -> Boolean
 * - isSitemapUserinfoChar(char, encode) -> Boolean
 * - isDomainChar(char, { start, end } = {}) -> Boolean
 * - isPathChar(char, encode) -> Boolean
 * - isSitemapPathChar(char, encode) -> Boolean
 * - isQueryOrFragmentChar(char, encode) -> Boolean
 * - isSitemapQueryOrFragmentChar(char, encode) -> Boolean
 * - isPercentEncodingChar(char) -> Boolean
 */
const { object: { is } } = require('../helpers');

/**
 * @func isSchemeChar
 *
 * Check scheme legal ascii codes according to
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.1.
 *
 * Scheme must start with a letter and be followed by any combination of
 * letters, digits, plus ("+"), period ("."), or hyphen ("-").
 *
 * Letters must be in lowercase.
 *
 * 43             +
 * 45             -
 * 46             .
 * 48 to 57       0-9
 * 97 to 122      a-z
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isSchemeChar = function isSchemeChar(char, { start } = {}) {
  if (!is(String, char)) {
    return false;
  }

  const code = char.charCodeAt();

  if (start) {
    return code >= 97 && code <= 122;
  }

  return (code >= 48 && code <= 57)
    || (code >= 97 && code <= 122)
    || code === 43
    || code === 45
    || code === 46;
};

/**
 * @func isUserinfoChar
 *
 * Check userinfo legal ascii codes according to
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.2.1.
 *
 * 33             !
 * 36             $
 * 37             % (not allowed when encoding)
 * 38 to 46       &'()*+,-.
 * 48 to 57       0-9
 * 58             :
 * 59             ;
 * 61             =
 * 65 to 90       A-Z
 * 95             _
 * 97 to 122      a-z
 * 126            ~
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isUserinfoChar = function isUserinfoChar(char, encode) {
  if (!is(String, char)) {
    return false;
  }

  // false by default
  const encoding = encode === true;
  const code = char.charCodeAt();

  // % is not allowed when encoding
  if (code === 37) {
    return !encoding;
  }

  return (code >= 38 && code <= 46)
    || (code >= 48 && code <= 57)
    || (code >= 65 && code <= 90)
    || (code >= 97 && code <= 122)
    || code === 33
    || code === 36
    || code === 58
    || code === 59
    || code === 61
    || code === 95
    || code === 126;
};

/**
 * @func isSitemapUserinfoChar
 *
 * Check sitemap userinfo legal ascii codes according to
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.2.1;
 * - https://support.google.com/webmasters/answer/183668?hl=en&ref_topic=4581190.
 *
 * Removed AZ ' and *
 *
 * 33             !
 * 36             $
 * 37             % (not allowed when encoding)
 * 38             & (allowed but must be a proper escape code)
 * 40 to 41       ()
 * 43 to 46       +,-.
 * 48 to 57       0-9
 * 58             :
 * 59             ;
 * 61             =
 * 95             _
 * 97 to 122      a-z
 * 126            ~
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isSitemapUserinfoChar = function isSitemapUserinfoChar(char, encode) {
  if (!is(String, char)) {
    return false;
  }

  // false by default
  const encoding = encode === true;
  const code = char.charCodeAt();

  // % is not allowed when encoding
  if (code === 37) {
    return !encoding;
  }

  return (code >= 40 && code <= 41)
    || (code >= 43 && code <= 46)
    || (code >= 48 && code <= 57)
    || (code >= 97 && code <= 122)
    || code === 33
    || code === 36
    || code === 38
    || code === 58
    || code === 59
    || code === 61
    || code === 95
    || code === 126;
};

/**
 * @func isDomainChar
 *
 * Check domain legal codes according to
 * - RFC-1034 https://www.ietf.org/rfc/rfc1034.txt.
 *
 * 45             -
 * 48 to 57       0-9
 * 97 to 122      a-z
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isDomainChar = function isDomainChar(char, { start, end } = {}) {
  if (!is(String, char)) {
    return false;
  }

  const code = char.charCodeAt();

  if ((start === true || end === true) && code === 45) {
    return false;
  }

  return (code >= 48 && code <= 57)
    || (code >= 97 && code <= 122)
    || code === 45;
};

/**
 * @func isPathChar
 *
 * Check path legal ascii codes according to
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.3.
 *
 * 33             !
 * 36 to 59       $%&'()*+,-./0-9:;
 * 61             =
 * 64 to 90       @A-Z
 * 95             _
 * 97 to 122      a-z
 * 126            ~
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isPathChar = function isPathChar(char, encode) {
  if (!is(String, char)) {
    return false;
  }

  // false by default
  const encoding = encode === true;
  const code = char.charCodeAt();

  // % is not allowed when encoding
  if (code === 37) {
    return !encoding;
  }

  return (code >= 36 && code <= 59)
    || (code >= 64 && code <= 90)
    || (code >= 97 && code <= 122)
    || code === 33
    || code === 61
    || code === 95
    || code === 126;
};

/**
 * @func isSitemapPathChar
 *
 * Check path legal ascii codes according to
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.3.
 *
 * Removed AZ ' and *
 *
 * 33             !
 * 36 to 38       $%& (& allowed but must be a proper escape code)
 * 40 to 41       ()
 * 43 to 59       +,-./0-9:;
 * 61             =
 * 64             @
 * 95             _
 * 97 to 122      a-z
 * 126            ~
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isSitemapPathChar = function isSitemapPathChar(char, encode) {
  if (!is(String, char)) {
    return false;
  }

  // false by default
  const encoding = encode === true;
  const code = char.charCodeAt();

  // % is not allowed when encoding
  if (code === 37) {
    return !encoding;
  }

  return (code >= 36 && code <= 38)
    || (code >= 40 && code <= 41)
    || (code >= 43 && code <= 59)
    || (code >= 97 && code <= 122)
    || code === 33
    || code === 61
    || code === 64
    || code === 95
    || code === 126;
};

/**
 * @func isQueryOrFragmentChar
 *
 * Check query/fragment legal ascii codes according to
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.4;
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.5.
 *
 * path char
 * 63             ?
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isQueryOrFragmentChar = function isQueryOrFragmentChar(char, encode) {
  if (isPathChar(char, encode)) {
    return true;
  }

  return is(String, char) && char.charCodeAt() === 63;
};

/**
 * @func isSitemapQueryOrFragmentChar
 *
 * Check query/fragment legal ascii codes according to
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.4;
 * - RFC-3986 https://tools.ietf.org/html/rfc3986#section-3.5.
 *
 * sitemap path char
 * 63             ?
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isSitemapQueryOrFragmentChar = function isSitemapQueryOrFragmentChar(char, encode) {
  if (isSitemapPathChar(char, encode)) {
    return true;
  }

  return is(String, char) && char.charCodeAt() === 63;
};

/**
 * @func isPercentEncodingChar
 *
 * Check percent encoding legal ascii codes according to RFC-3986 https://tools.ietf.org/html/rfc3986#section-2.1.
 *
 * 48 to 57       0-9
 * 65 to 70       A-F
 *
 * @param  {String} char
 * @return {Boolean}
 */
const isPercentEncodingChar = function isPercentEncodingChar(char) {
  if (!is(String, char)) {
    return false;
  }

  const code = char.charCodeAt();

  return (code >= 48 && code <= 57)
    || (code >= 65 && code <= 70);
};

module.exports = Object.freeze({
  isSchemeChar,
  isUserinfoChar,
  isSitemapUserinfoChar,
  isDomainChar,
  isPathChar,
  isSitemapPathChar,
  isQueryOrFragmentChar,
  isSitemapQueryOrFragmentChar,
  isPercentEncodingChar,
});
